apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/PublisherClient.cs

```diff
@@ -585,54 +754,82 @@ private void TriggerSend()
-            var batch = _batchesReady.Dequeue();
+            var readyBatch = _batchesReady.Dequeue();
+            var batch = readyBatch.Batch;
+            var state = readyBatch.State;
             _batchesInFlightCount += 1;
             // Update flow-control counts.
             _queueElementCount -= batch.Messages.Count;
             _queueByteCount -= batch.ByteCount;
+            // Send the batch
+            _taskHelper.Run(Send);
 
             async Task Send()
             {
                 // Perform the RPC to server, catching exceptions.
                 var publishTask = client.PublishAsync(TopicName, batch.Messages, CallSettings.FromCancellationToken(_hardStopCts.Token));
                 var response = await _taskHelper.ConfigureAwaitHideErrors(() => publishTask, null);
-                // Propagate task result to batch.
-                switch (publishTask.Status)
-                {
-                    case TaskStatus.RanToCompletion:
-                        batch.BatchCompletion.SetResult(response.MessageIds);
-                        break;
-                    case TaskStatus.Canceled:
-                        batch.BatchCompletion.SetCanceled();
-                        break;
-                    case TaskStatus.Faulted:
-                        batch.BatchCompletion.SetException(publishTask.Exception.InnerExceptions);
-                        break;
-                    default:
-                        throw new InvalidOperationException("Invalid TaskStatus");
-                }
-                // A client is now idle. Record it, and see if a further batch is ready to send.
+                Action postLockAction;
                 lock (_lock)
                 {
                     _batchesInFlightCount -= 1;
                     _idleClients.Enqueue(client);
                     ShutdownIfCompleted();
-                    if (_batchesReady.Count > 0)
+                    state.SetState(OrderingKeyState.Normal);
+                    var batches = state.Batches;
+                    switch (publishTask.Status)
                     {
-                        // Already a batch in the ready-queue, so just send it.
-                        TriggerSend();
+                        case TaskStatus.RanToCompletion:
+                            postLockAction = () => batch.BatchCompletion.SetResult(response.MessageIds);
+                            break;
+                        case TaskStatus.Canceled:
+                            postLockAction = () => batch.BatchCompletion.SetCanceled();
+                            break;
+                        case TaskStatus.Faulted:
+                            if (state.HasOrderingKey)
+                            {
+                                // For batches with an ordering-key: Retry transient errors forever;
+                                // otherwise fail all unsent messages with the same ordering-key, then refuse any more until `ResumePublish()` is called.
```

Comment:
The semi-colon in the first line is confusing, it makes me expect that the second line is about messages *without* an ordering key, but presumably it's about messages with an ordering key, but with a permanent error.
What about messages without an ordering key? That's covered later, but I think a single comment *above* the if statement that checks for the ordering key would be better.

Reply:
Clarified.